Sources/Sentry/SentryANRTracker.m

```diff
@@ -47,60 +48,34 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
 - (void)detectANRs
 {
     NSThread.currentThread.name = @"io.sentry.app-hang-tracker";
-
     self.thread = NSThread.currentThread;
 
-    BOOL wasPreviousANR = NO;
+    __block NSInteger ticksSinceUiUpdate = 0;
+    NSInteger reportTreshold = 5;
+    NSTimeInterval sleepInterval = self.timeoutInterval / reportTreshold;
 
     while (![self.thread isCancelled]) {
+        ticksSinceUiUpdate++;
 
-        NSDate *blockDeadline =
-            [[self.currentDate date] dateByAddingTimeInterval:self.timeoutInterval];
+        [self.dispatchQueueWrapper dispatchOnMainQueue:^{
+            ticksSinceUiUpdate = 0;
+            self->reported = NO;
+        }];
 
-        __block BOOL blockExecutedOnMainThread = NO;
-        [self.dispatchQueueWrapper dispatchOnMainQueue:^{ blockExecutedOnMainThread = YES; }];
+        [self.threadWrapper sleepForTimeInterval:sleepInterval];
 
-        [self.threadWrapper sleepForTimeInterval:self.timeoutInterval];
+        if (ticksSinceUiUpdate >= reportTreshold && !reported) {
+            self->reported = YES;
 
-        if (blockExecutedOnMainThread) {
-            if (wasPreviousANR) {
-                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelWarning];
-                [self ANRStopped];
+            if (![self.crashWrapper isApplicationInForeground]) {
+                [SentryLog logWithMessage:@"Ignoring ANR because the app is in the background"
+                                 andLevel:kSentryLevelDebug];
+                continue;
             }
 
-            wasPreviousANR = NO;
-            continue;
-        }
-
-        if (wasPreviousANR) {
-            [SentryLog logWithMessage:@"Ignoring ANR because ANR is still ongoing."
-                             andLevel:kSentryLevelDebug];
-            continue;
-        }
-
-        // The blockDeadline should be roughly executed after the timeoutInterval even if there is
-        // an ANR. If the app gets suspended this thread could sleep and wake up again. To avoid
-        // false positives, we don't report ANRs if the delta is too big.
-        NSTimeInterval deltaFromNowToBlockDeadline =
-            [[self.currentDate date] timeIntervalSinceDate:blockDeadline];
-
-        if (deltaFromNowToBlockDeadline >= self.timeoutInterval) {
-            NSString *message =
-                [NSString stringWithFormat:@"Ignoring ANR because the delta is too big: %f.",
-                          deltaFromNowToBlockDeadline];
-            [SentryLog logWithMessage:message andLevel:kSentryLevelDebug];
-            continue;
-        }
```

Comment:
h: this check is also missing from the new code

Reply:
The unity code doesn't seem to have something like this either. What kind of check would you expect here? As soon as `ticksSinceUiUpdate` hits 5, we report the ANR, mark it as reported, and that's it. I guess we could check if `ticksSinceUiUpdate` is more than 10, and then ignore it?

Reply:
I had to add that logic back in, to make the `testAppSuspended_NoANR` test pass. And looking at the name of that test, seems important to keep that functionality :)